src/StreamJsonRpc/NerdbankMessagePackFormatter.cs

```diff
@@ -0,0 +1,2301 @@
+        public override void Write(ref NBMP.MessagePackWriter writer, in RawMessagePack value, SerializationContext context)
+        {
+            value.WriteRaw(ref writer);
+        }
+
+        public override JsonObject? GetJsonSchema(JsonSchemaContext context, ITypeShape typeShape)
+        {
+            return CreateUndocumentedSchema(typeof(RawMessagePackConverter));
+        }
+    }
+
+    private class ProgressConverterResolver
+    {
+        private readonly NerdbankMessagePackFormatter mainFormatter;
+
+        internal ProgressConverterResolver(NerdbankMessagePackFormatter formatter)
+        {
+            this.mainFormatter = formatter;
+        }
+
+        public MessagePackConverter<T> GetConverter<T>()
+        {
+            MessagePackConverter<T>? converter = default;
+
+            if (MessageFormatterProgressTracker.CanDeserialize(typeof(T)))
+            {
+                converter = new PreciseTypeConverter<T>(this.mainFormatter);
+            }
+            else if (MessageFormatterProgressTracker.CanSerialize(typeof(T)))
+            {
+                converter = new ProgressClientConverter<T>(this.mainFormatter);
+            }
+
+            // TODO: Improve Exception
+            return converter ?? throw new NotSupportedException();
+        }
+
+        /// <summary>
+        /// Converts an instance of <see cref="IProgress{T}"/> to a progress token.
+        /// </summary>
+        private class ProgressClientConverter<TClass> : MessagePackConverter<TClass>
+        {
+            private readonly NerdbankMessagePackFormatter formatter;
+
+            internal ProgressClientConverter(NerdbankMessagePackFormatter formatter)
+            {
+                this.formatter = formatter;
+            }
+
+            public override TClass Read(ref NBMP.MessagePackReader reader, SerializationContext context)
+            {
+                throw new NotSupportedException("This formatter only serializes IProgress<T> instances.");
+            }
+
+            public override void Write(ref NBMP.MessagePackWriter writer, in TClass? value, SerializationContext context)
+            {
+                if (value is null)
+                {
+                    writer.WriteNil();
+                }
+                else
+                {
+                    long progressId = this.formatter.FormatterProgressTracker.GetTokenForProgress(value);
+                    writer.Write(progressId);
+                }
+            }
+
+            public override JsonObject? GetJsonSchema(JsonSchemaContext context, ITypeShape typeShape)
+            {
+                return CreateUndocumentedSchema(typeof(ProgressClientConverter<TClass>));
+            }
+        }
+
+        /// <summary>
+        /// Converts a progress token to an <see cref="IProgress{T}"/> or an <see cref="IProgress{T}"/> into a token.
+        /// </summary>
+        private class PreciseTypeConverter<TClass> : MessagePackConverter<TClass>
```

Comment:
Should "Precise" be "Progress"?

Reply:
I copied the name from MessagePackFormatter and I'm not sure what Precise refers to. The other type of converter created by this resolver is ProgressClient.

Reply:
I can WriteRaw(MyMPString.MsgPck.Span) and then ReadString() just fine.